lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,558 @@
+	notifyDelay time.Duration
+	// time after which an active event is passed to syncthing-core
+	notifyTimeout         time.Duration
+	notifyTimer           *time.Timer
+	notifyTimerNeedsReset bool
+	notifyTimerResetChan  chan time.Duration
+	notifyChan            chan []string
+	// All detected and to be scanned events are stored in a tree like
+	// structure mimicking folders to keep count of events per directory.
+	rootEventDir     *eventDir
+	backendEventChan chan notify.EventInfo
+	inProgress       map[string]struct{}
+	cfg              *config.Wrapper
+	configUpdate     chan config.FolderConfiguration
+	stop             chan struct{}
+}
+
+type Service interface {
+	Serve()
+	Stop()
+	C() <-chan []string
+	UpdateIgnores(ignores *ignore.Matcher)
+	VerifyConfiguration(from, to config.Configuration) error
+	CommitConfiguration(from, to config.Configuration) bool
+	String() string
+}
+
+func New(id string, cfg *config.Wrapper, ignores *ignore.Matcher) (Service, error) {
+	fsWatcher := &watcher{
+		folderID:              id,
+		folderIgnores:         ignores,
+		folderIgnoresUpdate:   make(chan *ignore.Matcher),
+		notifyTimerNeedsReset: false,
+		notifyTimerResetChan:  make(chan time.Duration),
+		notifyChan:            make(chan []string),
+		rootEventDir:          newEventDir(".", nil),
+		backendEventChan:      make(chan notify.EventInfo, maxFiles),
+		inProgress:            make(map[string]struct{}),
+		cfg:                   cfg,
+		configUpdate:          make(chan config.FolderConfiguration),
+		stop:                  make(chan struct{}),
+	}
+	folderCfg, ok := cfg.Folder(id)
+	if !ok {
+		panic(fmt.Sprintf("bug: Folder %s does not exist", id))
+	}
+	fsWatcher.updateConfig(folderCfg)
+
+	if err := fsWatcher.setupBackend(); err != nil {
+		return nil, err
+	}
+
+	return fsWatcher, nil
+}
+
+func (w *watcher) setupBackend() error {
+	absShouldIgnore := func(absPath string) bool {
+		if !isSubpath(absPath, w.folderPath) {
+			return true
+		}
+		relPath, _ := filepath.Rel(w.folderPath, absPath)
+		return w.folderIgnores.ShouldIgnore(relPath)
+	}
+	if err := notify.WatchWithFilter(filepath.Join(w.folderPath, "..."), w.backendEventChan, absShouldIgnore, w.eventMask()); err != nil {
+		notify.Stop(w.backendEventChan)
+		close(w.backendEventChan)
+		if isWatchesTooFew(err) {
+			err = watchesLimitTooLowError(w.folderDescription)
+		}
+		return err
+	}
+	return nil
+}
+
+func (w *watcher) Serve() {
+	w.notifyTimer = time.NewTimer(w.notifyDelay)
+	defer w.notifyTimer.Stop()
+
+	inProgressItemSubscription := events.Default.Subscribe(events.ItemStarted | events.ItemFinished)
+
+	w.cfg.Subscribe(w)
+
+	for {
+		// Detect channel overflow
+		if len(w.backendEventChan) == maxFiles {
+		outer:
+			for {
+				select {
+				case <-w.backendEventChan:
+				default:
+					break outer
+				}
+			}
+			// Issue full rescan as events were lost
+			w.newEvent(w.folderPath, nonRemove)
+			l.Debugln(w, "Backend channel overflow: Scan entire folder")
```

Comment:
The comment is misleading: The scan is not issued immediately, just an "artificial" event for the root dir is introduced which makes sure that a full rescan happens after the timeout. Meaning worst case scenario is a full rescan happening every `notifyTimeout` which is by default 1min. Good enough?

Also as long as this routine runs faster than the backend generates events, this should be able to scope with such moving scenarios by aggregating changes.